dbt/source.py

```diff
@@ -15,72 +17,91 @@ def __init__(self, project, own_project=None):
-        source_paths"""
-        found = []
-
-        if type(source_paths) not in (list, tuple):
-            source_paths = [source_paths]
-
-        for source_path in source_paths:
-            root_path = os.path.join(self.own_project_root, source_path)
-            for root, dirs, files in os.walk(root_path):
-                for filename in files:
-                    abs_path = os.path.join(root, filename)
-                    rel_path = os.path.relpath(abs_path, root_path)
-
-                    if fnmatch.fnmatch(filename, file_pattern):
-                        found.append(
-                            (self.project,
-                             source_path,
-                             rel_path,
-                             self.own_project)
-                        )
-        return found
+    def build_models_from_file_matches(
+            self,
+            to_build,
+            file_matches,
+            extra_args=[]):
+
+        build_args = [[self.project,
+                       file_match.get('searched_path'),
+                       file_match.get('relative_path'),
+                       self.own_project] + extra_args
+                      for file_match in file_matches]
+
+        return [to_build(*args) for args in build_args]
 
     def get_models(self, model_dirs, create_template):
-        pattern = "[!.#~]*.sql"
-        models = [Model(*model + (create_template,))
-                  for model in self.find(model_dirs, pattern)]
-        return models
+        file_matches = dbt.clients.system.find_matching(
+            self.own_project_root,
+            model_dirs,
+            "[!.#~]*.sql")
+
+        return self.build_models_from_file_matches(
+            Model,
+            file_matches,
+            [create_template])
 
     def get_test_models(self, model_dirs, create_template):
-        pattern = "[!.#~]*.sql"
-        models = [TestModel(*model + (create_template,))
-                  for model in self.find(model_dirs, pattern)]
-        return models
+        file_matches = dbt.clients.system.find_matching(
+            self.own_project_root,
+            model_dirs,
+            "[!.#~]*.sql")
+
+        return self.build_models_from_file_matches(
+            TestModel,
+            file_matches,
+            [create_template])
 
     def get_analyses(self, analysis_dirs):
-        pattern = "[!.#~]*.sql"
-        models = [Analysis(*analysis)
-                  for analysis in self.find(analysis_dirs, pattern)]
-        return models
-
-    def get_schemas(self, model_dirs):
-        "Get schema.yml files"
-        pattern = "[!.#~]*.yml"
-        schemas = [SchemaFile(*schema)
-                   for schema in self.find(model_dirs, pattern)]
-        return schemas
+        file_matches = dbt.clients.system.find_matching(
+            self.own_project_root,
+            analysis_dirs,
+            "[!.#~]*.sql")
+
+        return self.build_models_from_file_matches(
```

Comment:
i think we can clean it up even more if we drop the extreme OOP-iness in `models.py` but this works for now